presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushTopNThroughOuterJoinRuleSet.java

```diff
@@ -0,0 +1,220 @@
+ *          - TopN (present if Join is right or outer, not already limited, and orderBy symbols come from right source)
+ *             - right source
+ * </pre>
+ */
+public final class PushTopNThroughOuterJoinRuleSet
+{
+    private PushTopNThroughOuterJoinRuleSet() {}
+
+    public static Set<Rule<?>> rules()
+    {
+        return ImmutableSet.of(withProject(), withoutProject());
+    }
+
+    @VisibleForTesting
+    static Rule<?> withProject()
+    {
+        return new PushTopNThroughOuterJoinWithProject();
+    }
+
+    @VisibleForTesting
+    static Rule<?> withoutProject()
+    {
+        return new PushTopNThroughOuterJoinWithoutProject();
+    }
+
+    private static class PushTopNThroughOuterJoinWithProject
+            implements Rule<TopNNode>
+    {
+        private static final Capture<ProjectNode> PROJECT_CHILD = newCapture();
+        private static final Capture<JoinNode> JOIN_CHILD = newCapture();
+
+        private static final Pattern<TopNNode> PATTERN_WITH_PROJECT =
+                topN().with(source().matching(
+                        project().capturedAs(PROJECT_CHILD).with(source().matching(
+                                join().capturedAs(JOIN_CHILD).with(type().matching(type -> isLeftOrFullOuter(type) || isRightOrFullOuter(type)))))));
+
+        @Override
+        public Pattern<TopNNode> getPattern()
+        {
+            return PATTERN_WITH_PROJECT;
+        }
+
+        @Override
+        public Result apply(TopNNode parent, Captures captures, Context context)
+        {
+            ProjectNode projectNode = captures.get(PROJECT_CHILD);
+            JoinNode joinNode = captures.get(JOIN_CHILD);
+
+            Optional<SymbolMapper> symbolMapper = symbolMapper(parent.getOrderingScheme().getOrderBy(), projectNode.getAssignments());
+            if (!symbolMapper.isPresent()) {
+                return Result.empty();
+            }
+
+            TopNNode mappedTopN = symbolMapper.get().map(parent, joinNode, context.getIdAllocator().getNextId()).withStep(PARTIAL);
+
+            Optional<JoinNode> joinNodeWithTopNSource = joinNodeWithTopNSource(mappedTopN, joinNode, context);
+            return joinNodeWithTopNSource
+                    .map(newJoinNode ->
+                            Result.ofPlanNode(
+                                    parent.replaceChildren(ImmutableList.of(
+                                            projectNode.replaceChildren(ImmutableList.of(
+                                                    newJoinNode))))))
+                    .orElseGet(Result::empty);
+        }
+
+        private Optional<SymbolMapper> symbolMapper(List<Symbol> symbols, Assignments assignments)
+        {
+            SymbolMapper.Builder mapper = SymbolMapper.builder();
+            for (Symbol symbol : symbols) {
+                Expression expression = assignments.get(symbol);
+                if (!(expression instanceof SymbolReference)) {
+                    return Optional.empty();
```

Comment:
This could be checked in the pattern.

Reply:
How do I do this?

Reply:
Ah... `symbols` comes from TopN, so it is not possible today. Nevermind this comment.